### PR TITLE
Core/Formula Separation Step 1

### PR DESCRIPTION
**Pull request by xu-cheng**
This is step 1 for core/formula separation. i.e. moving formulae to `Homebrew/homebrew-core`. We will move core code to `Homebrew/brew` in step 2.

### CODE REVIEW

File: Library/Homebrew/cmd/update-report.rb [hunk 1]
```diff
@@ -6,26 +5,20 @@
...
+    if initial_revision.empty? || current_revision.empty?
       odie "update-report should not be called directly!"
     end
 
-    if master_reporter.updated?
-      initial_short = shorten_revision(master_reporter.initial_revision)
-      current_short = shorten_revision(master_reporter.current_revision)
-      puts "Updated Homebrew from #{initial_short} to #{current_short}."
-      hub.add(master_reporter)
+    if brew_updated = initial_revision != current_revision
```
Comment (mikemcquaid) [inline Library/Homebrew/cmd/update-report.rb]: Looks like `brew_updated` is unused?

Comment (mikemcquaid) [inline Library/Homebrew/cmd/update-report.rb]: Oh, it's used below. I'd probably assign it outside of the `if` in that case.

Comment (mikemcquaid) [inline Library/Homebrew/cmd/update-report.rb]: Until we have a `brew` repo separation the `brew` name here also isn't super clear to me.

File: Library/Homebrew/formula.rb [hunk 2]
```diff
@@ -1073,13 +1073,13 @@ def self.core_files
   # an array of all tap {Formula} names
   # @private
   def self.tap_names
-    @tap_names ||= Tap.flat_map(&:formula_names).sort
+    @tap_names ||= Tap.reject(&:core_tap?).flat_map(&:formula_names).sort
```
Comment (mikemcquaid) [inline Library/Homebrew/formula.rb]: Maybe worth making the `reject(&:core_tap?)` a method on the Tap class?

File: Library/Homebrew/tab.rb [hunk 3]
```diff
@@ -52,8 +52,9 @@ def self.from_file_content(content, path)
       attributes["source"]["tap"] = attributes.delete("tapped_from")
     end
 
-    if attributes["source"]["tap"] == "mxcl/master"
-      attributes["source"]["tap"] = "Homebrew/homebrew"
+    if attributes["source"]["tap"] == "mxcl/master" ||
+      attributes["source"]["tap"] == "Homebrew/homebrew"
+      attributes["source"]["tap"] = "Homebrew/core"
```
Comment (mikemcquaid) [inline Library/Homebrew/tab.rb]: Hopefully the last time we'll add to this :grinning:

File: Library/Homebrew/tap.rb [hunk 4]
```diff
@@ -446,23 +444,35 @@ def alias_file_to_name(file)
...
-# A specialized {Tap} class to mimic the core formula file system, which shares many
-# similarities with normal {Tap}.
-# TODO Separate core formulae with core codes. See discussion below for future plan:
-#      https://github.com/Homebrew/homebrew/pull/46735#discussion_r46820565
+# A specialized {Tap} class for the core formulae
 class CoreTap < Tap
+  if OS.mac?
+    OFFICIAL_REMOTE = "https://github.com/Homebrew/homebrew-core"
+  else
+    OFFICIAL_REMOTE = "https://github.com/Linuxbrew/homebrew-core"
```
Comment (mikemcquaid) [inline Library/Homebrew/tap.rb]: Does this exist yet?

File: Library/Homebrew/cmd/update-report.rb (hunk 1 above)
Comment (xu-cheng) [inline Library/Homebrew/cmd/update-report.rb]: What name do you suggest?

**Comment by mikemcquaid** [comment]
Makes sense to me so far. Is it not necessary for us to `rm -rf Library/Formula` too?

File: Library/Homebrew/formula.rb (hunk 2 above)
Comment (xu-cheng) [inline Library/Homebrew/formula.rb]: These methods are there to keep backward compatibility. I plan to deprecate `Fromula.tap_*` methods in the future.

File: Library/Homebrew/tap.rb (hunk 4 above)
Comment (xu-cheng) [inline Library/Homebrew/tap.rb]: It exists now.

**Comment by xu-cheng** [comment]
> Is it not necessary for us to rm -rf Library/Formula too?

No, we want to keep `Library/Formula` for the time being until we move to new `Homebrew/brew` repo. At then, it will be replaced as symlink.

File: Library/Homebrew/cmd/update-report.rb (hunk 1 above)
Comment (xu-cheng) [inline Library/Homebrew/cmd/update-report.rb]: Update to avoid this.

**Comment by UniqMartin** [comment]
> No, we want to keep `Library/Formula` for the time being until we move to new `Homebrew/brew` repo.

So does that mean that this PR will switch to using the new `core` tap and all of Homebrew will use the files from there and both `Library/Formula` and `Library/Aliases` (that are still retained) are simply ignored by all of Homebrew? And only after the second step we'll remove/replace them?

> At then, it will be replaced as symlink.

Do we really need that symlink? What would break (as far as is known) if we just wiped `Library/Formula` and didn't provide the symlink? I'm just thinking that not providing it in the first place might be the cleaner solution and create a higher incentive for those who have this path hard-coded to fix their tools.

**Comment by xu-cheng** [comment]
> both Library/Formula and Library/Aliases (that are still retained) are simply ignored by all of Homebrew

Yes, both of them will be simply ignored.

> Do we really need that symlink? What would break (as far as is known) if we just wiped Library/Formula and didn't provide the symlink?

The major problem is some script used by users may hard code the path of `Library/Formula`.

**Comment by UniqMartin** [comment]
> The major problem is some script used by users may hard code the path of `Library/Formula`.

What I'm thinking is that it isn't unlikely the same script is also hard-coding `Library/Taps/*/*` and blindly iterating over all the entries it will find. That would mean that it will see all the core formulae twice (once via the `Taps` directory and once via the symlinked legacy location) and that could also create some strange issues. If `Library/Formula` was simply gone, the cause of the breakage could be more obvious.

But well, it's hard to guess what scripts might be floating around and what assumptions they make, so quite likely I'm totally wrong on that one. :smiley_cat:

**Comment by mikemcquaid** [comment]
> Do we really need that symlink? What would break (as far as is known) if we just wiped Library/Formula and didn't provide the symlink? I'm just thinking that not providing it in the first place might be the cleaner solution and create a higher incentive for those who have this path hard-coded to fix their tools.

I'd like us to include it at first and consider removing it in future.

**Comment by mikemcquaid** [comment]
> What I'm thinking is that it isn't unlikely the same script is also hard-coding Library/Taps/*/* and blindly iterating over all the entries it will find. That would mean that it will see all the core formulae twice (once via the Taps directory and once via the symlinked legacy location) and that could also create some strange issues.

While this concerns me a bit too I'd rather prioritise a slower user migration over not breaking these scripts (even if ideally we'd do neither).

**Comment by UniqMartin** [comment]
I might be severely off due to my much more limited experience with the project (or more generally any project of that scale), but my gut feeling is that by being too careful not to break other people's code and workflows, we're also amassing a certain amount of cruft in our own code, which is not benefiting us in the mid- to long-term.

I guess there's a sweet spot between being overly reckless and not being bold enough. I'm not sure where it is, but I'd like us to think about this (which shall not imply that we're not already doing that). So in the likely case that I'm draining resources without getting anywhere, feel free to tell me to shut up. :see_no_evil:

**Comment by mikemcquaid** [comment]
> I might be severely off due to my much more limited experience with the project (or more generally any project of that scale), but my gut feeling is that by being too careful not to break other people's code and workflows, we're also amassing a certain amount of cruft in our own code, which is not benefiting us in the mid- to long-term.

This is definitely a consideration but it's something users don't care about at all. Cruftiness matters only to maintainers as it's something that can slow us down from developing e.g. new features but I'd definitely err on the side of incremental changes for migrations as dramatic as this.

**Comment by xu-cheng** [comment]
Is there any code comment for this PR? Oh when is the time to ship this?

**Comment by mikemcquaid** [comment]
@xu-cheng I'm happy with the code and :ok: for you to ship this whenever you're happy with it. Please feel free to rope in others on testing this, though.

**Comment by xu-cheng** [comment]
cc @Homebrew/maintainers for testing.

If all goes well, I plan to ship this in the weekend so I could help to fix unexpected bugs if there is any.

A side question, is there anyone welling to work on document change? e.g. README for `Homebrew/core` etc.

**Comment by mikemcquaid** [comment]
@xu-cheng I can help with document changes but won't be able to do so until the weekend. Can you detail what you'd like us to test, how we can verify it's working/broken and, ideally, a step-by-step guide to do so?

Also, just so I'm definitely sure I understand here: after we merge this we will no longer accept formula PRs on this repository, right?

**Comment by bfontaine** [comment]
Do we need to change the installer too or will it work as is for now?

**Comment by xu-cheng** [comment]
>  Can you detail what you'd like us to test, how we can verify it's working/broken and, ideally, a step-by-step guide to do so?

This PR doesn't change much on update process. And new core tap will be auto installed whenever necessary. So to test it, you can just pull this change locally and run whatever commands in your mind.

> Also, just so I'm definitely sure I understand here: after we merge this we will no longer accept formula PRs on this repository, right?

That's right. We will no longer accept formula PRs on this repo. In addition, I created a `brew legacy-pull`([gist](https://gist.github.com/xu-cheng/545163571f6c6ae74d6a)) as a temporary command to allow us to pull existing PR from this repo to `Homebrew/core`. 

> Do we need to change the installer too or will it work as is for now?

We will update the installer by adding one line code `brew tap homebrew/core`. But it should work without any change as well.

**Comment by mikemcquaid** [comment]
@xu-cheng Cool. Would be good to also queue up installer PRs and add a `--legacy` argument or autodetect from the URL when pulling from this repo to allow us to avoid duplicaion.

**Comment by mikemcquaid** [comment]
I think we need to figure out (and document) what we're going to do with future PRs/issues on this repository before we do this migration.

**Comment by xu-cheng** [comment]
I think the we have no problem for existed issues, since all of them can be easily to be relocated using https://github-issue-mover.appspot.com

For existed PRs, we need to treat them differently:
* PR on core code change: nothing changed.
* PR on both core code change and formulae change: PR authors have to resubmit the PR.
* PR on formulae change:
   * CI test is passed and PR is ready to be pulled: We can pull them using `brew legacy-pull`.
   * Otherwise: PR authors have to resubmit the PR.

**Comment by DomT4** [comment]
> I think the we have no problem for existed issues, since all of them can be easily to be relocated using https://github-issue-mover.appspot.com

I'd really like us to try _not_ to use that tool en mass if there's any workable alternative. As useful as it is for short issues, for longer ones with discussion it means everyone in the thread gets pinged into a new thread all over again without any actual progress being made. Was wondering if Mike has special powers to move the Issues in a cleaner way.

**Comment by mikemcquaid** [comment]
> I'd really like us to try not to use that tool en mass if there's any workable alternative.

Agreed.

I've not had a chance to test this this weekend and I think we need to wait longer and include e.g. documentation changes into this PR or we risk confusing a bunch of contributors.

**Comment by xu-cheng** [comment]
> I've not had a chance to test this this weekend and I think we need to wait longer and include e.g. documentation changes into this PR or we risk confusing a bunch of contributors.

Agreed. And could we have document changes on a separated PR? i.e. merge two PRs at the same time after everything is sorted out. So we won't have a PR having too many commits and file changes.

And preferably it would be great to have another maintainer to conduct document change. (read: I'm not good at writing document :cry:)

**Comment by mikemcquaid** [comment]
I think it would be good to have both on the same PR. Can you try the documentation changes and then I'll give you a hand with the language (or make a PR to your PR).

**Comment by mikemcquaid** [comment]
@xu-cheng How does next Saturday (2nd April) sound to merge this? I can be around all day UK time in case anything goes wrong. What documentation do you think we need to update before then? CC @Homebrew/maintainers for a change because this is a really big deal.

**Comment by apjanke** [comment]
I will also be around Saturday April 2nd during the day US time, from about noon to 9 PM Eastern Time.

**Comment by xu-cheng** [comment]
Yes, next weekend is good to me.

For the code, I don't know why `brew update-test` will hang out forever on bot. But I cannot reproduce it locally. 

For the document, it's mainly about how PR authors contribute to new core tap. I hope I can come a draft PR on this before next weekend. But I'm a little occupied recently.

**Comment by bfontaine** [comment]
I haven’t been able to contribute much time to Homebrew recently but that should be better in a couple of weeks. I should be around next Saturday too.

**Pull request closed without merging**